### ISSUE TITLE
[FEATURE]: Model Saving and Checkpointing

### ISSUE THREAD

**Issue by nmichlo** [nmichlo/disent#28]
**Is your feature request related to a problem? Please describe.**
Model saving and checkpointing is currently disabled for `experiment/run.py`
This was due to old pickling errors and the extensive use of wandb for logging. Actual saved models were not needed at the time.

**Describe the solution you'd like**
Re-enable model checkpointing, and allow continuing of training.
- Schedules will need to be pickled 
- Add config options to enable saving and continuing training

**Comment by meffmadd** [comment]
Hi 👋 First of all thanks for the fantastic framework!

I started to work on this issue since I need model saving for my visualization project. Model saving seems to be very simple, however, loading a checkpoint appears to be a bit more complicated as far as I understand. I found that a saved checkpoint can be loaded but the model parameter for e.g. `BetaVae` is an `omegaconf.dictconfig.DictConfig` instead of an actual `AutoEncoder` object. Is this how the models are serialized in general when omegaconf is available?

Anyhow, maybe a simpler solution would be to pass simple arguments (like `x_shape`) to `BetaVae` etc. so that the serialization is simpler. If you want I could implement this but this would of course introduce breaking changes to the API. Probably you also know a lot more about this than I do.

**Comment by meffmadd** [comment]
Ah, it seems like someone hit the same roadblock before: https://github.com/Lightning-AI/lightning/discussions/6144

**Comment by nmichlo** [comment]
Hi @meffmadd, really glad you are finding it useful!

I managed to get away with wandb results a while back so I never got around to fixing this.

I'll investigate based on the information you provided and get back to you. Thank you for that!

**Comment by nmichlo** [comment]
You noted the object is an OmegaConf instance. I don't think it would break too much if we switch that over to a dictionary, and recursively convert all the values. There is a built in function for this.

(As for pytorch lightning, I have become a bit disillusioned towards it, as it has placed certain constraints on the framework that were never intended.)

To get to my question, how important is API stability for you right now?

**Comment by meffmadd** [comment]
API stability it is not a major concern for me. I only planned on using the hydra configs but maybe using the Python API is more useful for me (since it removes a layer of complexity). Yeah, frameworks are nice as long as their magic works and there is documentation 😅

For model saving when running with `hydra` I think I found a simple workaround:

```python
    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
        checkpoint["hyper_parameters"]["model"] = self._model
        return super().on_save_checkpoint(checkpoint)
```

This manually sets the model in the checkpoint and also works for loading!

And a quick unrelated question:
Could you tell me a config for beta-VAE that works well with dSprites? I use `bce` loss but this somehow creates NaN values in the encoder output. If I use the norm_conv64 framework it works well but this seems to be not be standard as per your warning.

**Comment by nmichlo** [comment]
Thank you so much for the PR! I left a few comments about tests. We just need to make sure to add the new keys to the configs and (possibly) update the tests to tests the checkpointing.

As for your question. That should not be happening with the BCE loss. It may be due to largely unrelated things like the strength of the regularization term, or the learning rate too.
- I know dSprites is a binary dataset, but I would still recommend using MSE loss as the BCE loss has lead to prevalent errors in the VAE field. For example moving from `dsprites` to `dsprites-imagenet` or even `shapes3d`.

**Comment by meffmadd** [comment]
I will fix the configs now!

Thanks for your answer! I will try it with MSE but with a higher learning rate because when I tested the beta-VAE with MSE it did not converge at all.

**Comment by nmichlo** [comment]
I think possibly a lower beta value then too.

**Comment by nmichlo** [comment]
Closing this with your changed in #37, now released under #39
